bin/pytorch_inference/Main.cc

```diff
@@ -93,6 +83,16 @@ void writeTensor(const torch::TensorAccessor<T, 1UL>& accessor,
 template<typename T>
 void writeTensor(const torch::TensorAccessor<T, 2UL>& accessor,
                  ml::core::CRapidJsonLineWriter<rapidjson::OStreamWrapper>& jsonWriter) {
+    jsonWriter.StartArray();
+    for (int i = 0; i < accessor.size(0); ++i) {
+        writeTensor(accessor[i], jsonWriter);
+    }
+    jsonWriter.EndArray();
+}
+
+template<typename T>
+void writeTensor(const torch::TensorAccessor<T, 3UL>& accessor,
+                 ml::core::CRapidJsonLineWriter<rapidjson::OStreamWrapper>& jsonWriter) {
```

Comment:
This seems surprising to me: if you `accessor.size(0) == 1` then I think this method should assert that; otherwise this needs a `jsonWriter.StartArray();` and `jsonWriter.EndArray();` surely or else the JSON is malformed. This code could be made fully general and clearer IMO:
```
template<typename T, std::size_t N>
struct SDoWriteTensor {
    static void dispatch(const torch::TensorAccessor<T, N>& accessor,
                     ml::core::CRapidJsonLineWriter<rapidjson::OStreamWrapper>& jsonWriter) {
        jsonWriter.StartArray();
        for (int i = 0; i < accessor.size(0); ++i) {
            writeTensor(accessor[i])
        }
        jsonWriter.EndArray();
    }
};
template<typename T>
struct SDoWriteTensor<T, 1> {
    static void dispatch(const torch::TensorAccessor<T, 1>& accessor,
                     ml::core::CRapidJsonLineWriter<rapidjson::OStreamWrapper>& jsonWriter) {
        jsonWriter.Double(static_cast<double>(accessor[i]));
    }
};
template<typename T, std::size_t N>
void writeTensor(const torch::TensorAccessor<T, N>& accessor,
                     ml::core::CRapidJsonLineWriter<rapidjson::OStreamWrapper>& jsonWriter) {
    SDoWriteTensor<T, N>::dispatch(accessor, jsonWriter);
}
```
(Note as well looking at `TensorAccessor.h` it seems that `torch::TensorAccessor<T, 1>` is a scalar.) You have to wrap in a struct because partial template specialisation isn't allowed for functions, but I don't think this hurts readability much.

Reply:
The hack I was working around here is that in all cases I want the output to be a 3D array but some models/tasks give a 2D output e.g. sentiment analysis. The missing start and end array are actually written elsewhere and this meant that a 2D output was embedded in a 3rd dimension whereas the 3D output was not. I've added comments and rewritten to make it clear why this is happening.